Blog: Pulumi is imperative, declarative & imperative (#1923)
pulumi/pulumi-hugo: themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

- we allow for policies to be written with the same ease as abstractions.
- we mix it all in a single setup

Sketch the setup of the article as the description of the PR for reviewers
-->

On a regular basis, articles and tweets pass by discussing whether or not some specific tool is imperative or declarative.
It's no surprise that Pulumi is often the tool being debated. What if I tell you that Pulumi is imperative, declarative and imperative?

<!--more-->

When we look at our [Frequently Asked Questions](/docs/support/faq/#is-pulumi-imperative-or-declarative), we read the following on the declarative versus imperative topic:

> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program.
> It’s not used for talking to the cloud provider API.

The title of the article mentions `imperative` and `declarative` both. Pulumi leverages the best of both worlds
into our product.

{{< tweet id="1553431913691430912" user="adamhjk" >}}

Pulumi tries to offer a solution where our customers are only limited by their imagination rather than the tool at hand.

> Let me use the Pulumi architecture to highlight why I mention *imperative* twice.

Reviewer: Would it be a good idea to define `imperative` and `declaritve`?  Something like this:

`Both terms refer to how the user provides direction to the automation platform. With an imperative tool, you define the steps to execute in order to reach the desired solution. With a declarative tool, you define the desired state of the final solution, and the automation platform determines how to achieve that state.`

I got the above from the following [link](https://www.linode.com/blog/devops/declarative-vs-imperative-in-iac/)

## Pulumi Architecture

> If you don't know the Pulumi architecture, here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:

Reviewer: Replace

> If you don't know the Pulumi architecture, here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:

Reviewer: Strike Out:  `If you don't know the Pulumi architecture, here `. 
Have the sentence start with:  Here is the...

Author: Done

### Language Host

Under the hood, our Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in
the `Pulumi.yaml` project file. Here is a small Python example:

```python
from pulumi_aws import s3

bucket = s3.Bucket('bucket')

for i in range(10):
    s3.BucketObject(
        f'object-{i}',
        s3.BucketObjectArgs(
            bucket=bucket.id,
            key=str(i),
        )
    )
```

A Pulumi program models the to-be state of your infrastructure. If you read the program above, you can see that
we define 11 resources as our to be infrastructure:

* 1 AWS S3 bucket
* 10 Objects in the bucket created in the previous step

While this is definitely an imperative program, there is one important thing to understand: instantiating an
`s3.Bucket`, `s3.BucketObject` or any other Pulumi resource should not be interpreted as an imperative creation
of the resource in the language host. Behind the scenes, any resource instantiation in the language host triggers a
`Register Resource` request to the Pulumi engine.

No matter what the actual state of your infrastructure is, running your program will always send to intended

> resource model to the engine.

Reviewer: What is sent to the engine?  
Maybe re-order this sentence, it sounds confusing since it seems like we left out a word. 

`Running your program always sends the resource model to the Pulumi Engine regardless of what state your infrastructure is in` - you can reword it however you want.  I

Author: Done.

to compute which actions need to be executed to bring the actual state in line with the intended model.

Our little example contains dependencies:
every `s3.BucketObject` uses the `bucket.id` as way to define in which bucket these objects should be stored.
The property `id` from the `s3.Bucket` is an `Output`. Outputs are Pulumi's way of tracking which property of one resource
is required by another, hereby creating a dependency between the resources. Our engine uses all these outputs
passed from one resource to another as the vertices in a directed acyclic graph (DAG). The engine determines
the order of actions based on this graph.

On a first run of `pulumi up` of our example program, our engine will first create the bucket, wait for
the provisioning to be complete, after which the actual bucket id is passed to the creation of
all the bucket objects. Since none of the bucket objects depend on other resources, these bucket
objects can all be provisioned concurrently. The engine will

On a second run, assuming no modifications to our example program, the Pulumi engine will compare the to be model
with the actual state and conclude that nothing needs to be done.

> Let's crank up the nr of bucket objects to 11.

Reviewer: Replace `nr` 
With `number`

Author: Done.

```python
from pulumi_aws import s3

bucket = s3.Bucket('bucket')

for i in range(11): # <- nr of objects increased by 1
    s3.BucketObject(
        f'object-{i}',
        s3.BucketObjectArgs(
            bucket=bucket.id,
            key=str(i),
        )
    )
```

The third Pulumi run will find the bucket and the first 10 objects in the state. The only action taken

> now is the creation of the 11th bucket object. This brings the actual state back in sync with your

Reviewer: Would it be a good idea to capture the snippet from `pulumi preview` that shows that only the 11th object will be created?

Author: Done.

---
title: "Pulumi Is Imperative, Declarative and Imperative"
date: 2022-09-30T12:28:19+02:00
meta_desc: Pulumi is often debated to be imperative or declarative.
  What if I tell you that Pulumi is imperative, declarative and imperative?
meta_image: meta.png
authors:
    - ringo-de-smet
tags:
    - cloud-engineering
    - infrastructure-as-code
---

On a regular basis, articles and tweets pass by discussing whether or not some specific tool is imperative or declarative.
It's no surprise that Pulumi is often the tool being debated. What if I tell you that Pulumi is imperative, declarative and imperative?

<!--more-->

When we look at our [Frequently Asked Questions](/docs/support/faq/#is-pulumi-imperative-or-declarative), we read the following on the declarative versus imperative topic:

> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program.
> It’s not used for talking to the cloud provider API.

> It is good to know to refresh what `declarative` and `imperative` mean:

Reviewer 2: Also, I did a Quick Bites on imperative vs declarative infrastructure that got into this confusion (and how a tool can be both): https://www.youtube.com/watch?v=4B7vLfVYG84. It may be worth linking somewhere here.

> It is good to know to refresh what `declarative` and `imperative` mean:

Author: As the outcome of the conversation we have over Slack, I'll update the headings of the 3 Pulumi architectural components as you proposed:

* "An Imperative Part: The Language Host"
* "The Declarative Part: The CLI and Engine"
* "Another Imperative Part: The Providers"

Makes it indeed clearer why I titled the article: `Pulumi is imperative, declarative, and imperative`

Thanks for the help!

> Both terms refer to how the user provides direction to the automation platform. With an imperative tool, you define the steps to execute in order to reach the desired solution. With a declarative tool, you define the desired state of the final solution, and the automation platform determines how to achieve that state. ([Source](https://www.linode.com/blog/devops/declarative-vs-imperative-in-iac/))

The title of the article mentions `imperative` and `declarative` both. Pulumi leverages the best of both worlds
into our product.

{{< tweet id="1553431913691430912" user="adamhjk" >}}

Pulumi tries to offer a solution where our customers are only limited by their imagination rather than the tool at hand.
Let me use the Pulumi architecture to highlight why I mention *imperative* twice.

## Pulumi Architecture

> Here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:

Reviewer 2: ```suggestion
Here is the diagram from our [How Pulumi Works]({{< relref "/docs/intro/concepts/how-pulumi-works" >}}) page:
```

![Pulumi Architecture](/images/docs/reference/engine-block-diagram.png)

> You code your infrastructure in your preferred programming language. When you are done coding, you run `pulumi up` and the Pulumi CLI starts

Reviewer 2: Per style, we don't break at 120 characters in our markdown as it throws off some reviewers. https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md#paragraphs-and-line-breaks

the language host for your selected programming language, as well as the required providers. The interaction between these 3 parts of
the architecture results in the actual creation or modification of your infrastructure.

### Language Host

> Under the hood, the Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in

Reviewer 2: ```suggestion
Under the hood, the Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime, which is configured in
```

> the `Pulumi.yaml` project file. Here is a small Python example:

Reviewer 2: There's a slight disconnect here. I was scrolling hunting for an example of the configuration in the project file, and instead there's a program file. May I suggest the following change?
```suggestion
the `Pulumi.yaml` project file. Once the language runtime is up, it turns to the program file. Here is a small Python example of a program file:
```

```python
from pulumi_aws import s3

bucket = s3.Bucket('bucket')

for i in range(10):
    s3.BucketObject(
        f'object-{i}',
        s3.BucketObjectArgs(
            bucket=bucket.id,
            key=str(i),
        )
    )
```

> A Pulumi program models the to-be state of your infrastructure. If you read the program above, you can see that

Reviewer 2: We try to avoid directional language (e.g., "above") and sensory language (e.g., "see"). https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md#language

```suggestion
A Pulumi program models the to-be state of your infrastructure. If you read this program, you can find that
```

we define 11 resources as our to-be infrastructure:

> * 1 [AWS S3](https://www.pulumi.com/registry/packages/aws/api-docs/s3/bucket/) bucket

Reviewer 2: ```suggestion
* 1 [AWS S3]({{< relref "/registry/packages/aws/api-docs/s3/bucket" >}}) bucket
```

> * 10 [Objects](https://www.pulumi.com/registry/packages/aws/api-docs/s3/bucketobject/) in the bucket created in the previous step

Reviewer 2: ```suggestion
* 10 [Objects]({{< relref "/registry/packages/aws/api-docs/s3/bucketobject" >}}) in the bucket created in the previous step
```

> The stack itself is also modelled as a resource and it is the parent to all other resources.

Reviewer 2: ```suggestion
The stack itself is also modelled as a resource, and it is the parent to all other resources.
```

```sh
$ pulumi up

Updating (<masked>/dev)

View Live: https://app.pulumi.com/<masked>/blog_code/dev/updates/1

     Type                    Name           Status
 +   pulumi:pulumi:Stack     blog_code-dev  created
 +   ├─ aws:s3:Bucket        bucket         created
 +   ├─ aws:s3:BucketObject  object-0       created
 +   ├─ aws:s3:BucketObject  object-1       created
 +   ├─ aws:s3:BucketObject  object-3       created
 +   ├─ aws:s3:BucketObject  object-2       created
 +   ├─ aws:s3:BucketObject  object-6       created
 +   ├─ aws:s3:BucketObject  object-4       created
 +   ├─ aws:s3:BucketObject  object-5       created
 +   ├─ aws:s3:BucketObject  object-7       created
 +   ├─ aws:s3:BucketObject  object-9       created
 +   └─ aws:s3:BucketObject  object-8       created

Resources:
    + 12 created

Duration: 13s
```

> While this is definitely an imperative program, there is one important thing to understand: instantiating an

Reviewer 2: ```suggestion
While this is definitely an imperative program, there is one important thing to understand: Instantiating an
```

> `s3.Bucket`, `s3.BucketObject` or any other Pulumi resource should not be interpreted as an imperative creation

Reviewer 2: ```suggestion
`s3.Bucket`, `s3.BucketObject`, or any other Pulumi resource should not be interpreted as an imperative creation
```

of the resource in the language host. Behind the scenes, any resource instantiation in the language host triggers a

> `Register Resource` request to the Pulumi engine. All these `Register Resource` requests together form the

Reviewer 2: I'd use the same wording we use in the How Pulumi Works page specifically in case someone starts exploring further.

```suggestion
`Register Resource` request to the Pulumi deployment engine. All these `Register Resource` requests together form the
```

resource model you as an infrastructure developer want to get in the end.

> Running your program always sends the full resource model to the Pulumi Engine regardless of what state your

Reviewer 2: ```suggestion
Running your program always sends the full resource model to the Pulumi deployment engine regardless of what state your
```

current infrastructure is in.

> Our previous example was shown in Python, but recently we also delivered [support for YAML](https://www.pulumi.com/docs/intro/languages/yaml/).

Reviewer 2: ```suggestion
Our previous example was shown in Python, but recently we also delivered [support for YAML]({{< relref "/docs/intro/languages/yaml" >}}).
```

The creation of our S3 bucket could be converted to this snippet:

```yaml
resources:
  bucket:
    type: aws:s3:Bucket
```

The separation of language support from the engine allows us to offer both imperative & declarative solutions.
But remember that none of the supported languages do any of the provisioning directly in the language host.

If the resource provisioning is not taking place in the language host, where is the magic happening then?

### CLI and Engine

In the previous step, you found out that the language host sends requests to the engine to fullfil your to-be
infrastructure.

> It is now that the Pulumi engine gets to work. The engine combines the intended model of the infrastructure

Reviewer 2: ```suggestion
It is now that the Pulumi deployment engine gets to work. The engine combines the intended model of the infrastructure
```

> received from the language host, the current state recorded in the state backend and the actual resource state

Reviewer 2: ```suggestion
received from the language host, the current state recorded in the state backend, and the actual resource state
```

to compute which actions need to be executed to bring the actual state in line with the intended model.

Our little example contains dependencies:

> every `s3.BucketObject` uses the `bucket.id` as way to define in which bucket these objects should be stored.

Reviewer 2: ```suggestion
Every `s3.BucketObject` uses the `bucket.id` as way to define in which bucket these objects should be stored.
```

The property `id` from the `s3.Bucket` is an `Output`. Outputs are Pulumi's way of tracking which property of one resource
is required by another, hereby creating a dependency between the resources. The engine uses all these outputs

> passed from one resource to another as the vertices in a directed acyclic graph (DAG). The engine determines

Reviewer 2: May I recommend linking to something that explains DAGs? A number of folks won't know what that means.

the order of actions based on this graph.

> On a first run of `pulumi up` of our example program, the engine will first create the bucket, wait for

Reviewer 2: ```suggestion
On a first run of `pulumi up` of our example program, the engine will first create the bucket and then wait for
```

> the provisioning to be complete, after which the actual bucket id is passed to the creation of

Reviewer 2: ```suggestion
the provisioning to be complete, after which the actual bucket ID is passed to the creation of
```

all the bucket objects. Since none of the bucket objects depend on other resources, these bucket
objects can all be provisioned concurrently.

On a second run, assuming no modifications to our example program, the Pulumi engine will compare the to-be model
with the actual state and conclude that nothing needs to be done.

Let's crank up the number of bucket objects to 11.

```python
from pulumi_aws import s3

bucket = s3.Bucket('bucket')

> for i in range(11): # <- nr of objects increased by 1

Reviewer 2: ```suggestion
for i in range(11): # <- number of objects increased by 1
```

    s3.BucketObject(
        f'object-{i}',
        s3.BucketObjectArgs(
            bucket=bucket.id,
            key=str(i),
        )
    )
```

The third Pulumi run will find the bucket and the first 10 objects in the state. The only action taken
now is the creation of the 11th bucket object. This brings the actual state back in sync with your
intended model.

```sh
$ pulumi up

Updating (<masked>/dev)

View Live: https://app.pulumi.com/<masked>/blog_code/dev/updates/2

     Type                    Name           Status
     pulumi:pulumi:Stack     blog_code-dev
 +   └─ aws:s3:BucketObject  object-10      created

Resources:
    + 1 created
    12 unchanged

Duration: 5s

```

Although you create the intended model of your infrastructure with an imperative language, the engine definitely
processes this in a declarative way.

### Providers

> The above examples hopefully made clear that the Pulumi engine calculates a set of actions to declaratively bring

Reviewer 2: ```suggestion
These examples hopefully made clear that the Pulumi engine calculates a set of actions to declaratively bring
```

the actual state in sync with your intended model. But the engine is not the component which knows how to
talk to all the different APIs of cloud and tool vendors. That's the role of the provider.

> The engine and the provider processes are connected with a gRPC connection. This is similar to the connection

Reviewer 2: I'd link to something about gRPC here.

between the language host and the engine, with the only difference of which requests are sent. The API
that providers expose are of a [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) nature:

> * Create: create a new resource
> * Read: read information of an existing resource
> * Update: update an existing resource with modified information
> * Delete: delete an existing resource when no longer needed

Reviewer 2: ```suggestion
* _Create_: create a new resource
* _Read_: read information of an existing resource
* _Update_: update an existing resource with modified information
* _Delete_: delete an existing resource when no longer needed
```

The providers receive requests from the engine dependent on the set of actions the engine calculated.

> A provider doesn't know anything about the state, the correlation between resource and so forth.

Reviewer 2: ```suggestion
A provider doesn't know anything about the state, the correlation between resource, and so forth.
```

The nature of the provider API is clearly imperative.

### Summary

If the question ever pops up again whether Pulumi is declarative or imperative, the answer is clearly we are both.
It is only based on which component of our architecture you are talking about:

> * Language host: imperative (JS/TS, Go, Python) and declarative (YAML)
> * Pulumi engine: declarative
> * Providers: imperative

Reviewer 2: ```suggestion
* _Language host_: imperative (JS/TS, Go, Python, C#/F#/.NET, Java) and declarative (YAML)
* _Pulumi deployment engine_: declarative
* _Providers_: imperative
```

To us, it mainly matters if we can solve your infrastructure automation problem. If you still miss something

> in our offering, we want to [hear from you](https://www.pulumi.com/contact/)!

Reviewer 2: ```suggestion
in our offering, we want to [hear from you]({{< relref "/contact" >}})!
```

The title of the article mentions `imperative` and `declarative` both. Pulumi leverages the best of both worlds
into our product.

{{< tweet id="1553431913691430912" user="adamhjk" >}}

Pulumi tries to offer a solution where our customers are only limited by their imagination rather than the tool at hand.
Let me use the Pulumi architecture to highlight why I mention *imperative* twice.

## Pulumi Architecture

Here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:

![Pulumi Architecture](/images/docs/reference/engine-block-diagram.png)

You code your infrastructure in your preferred programming language. When you are done coding, you run `pulumi up` and the Pulumi CLI starts
the language host for your selected programming language, as well as the required providers. The interaction between these 3 parts of
the architecture results in the actual creation or modification of your infrastructure.

### Language Host

> Under the hood, our Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in

Reviewer 3: ```suggestion
Under the hood, the Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in
```

`s3.Bucket`, `s3.BucketObject` or any other Pulumi resource should not be interpreted as an imperative creation
of the resource in the language host. Behind the scenes, any resource instantiation in the language host triggers a
`Register Resource` request to the Pulumi engine. All these `Register Resource` requests together form the
resource model you as an infrastructure developer want to get in the end.

Running your program always sends the full resource model to the Pulumi Engine regardless of what state your
current infrastructure is in.

Our previous example was shown in Python, but recently we also delivered [support for YAML](https://www.pulumi.com/docs/intro/languages/yaml/).
The creation of our S3 bucket could be converted to this snippet:

```yaml
resources:
  bucket:
    type: aws:s3:Bucket
```

The separation of language support from the engine allows us to offer both imperative & declarative solutions.
But remember that none of the supported languages do any of the provisioning directly in the language host.

If the resource provisioning is not taking place in the language host, where is the magic happening then?

### CLI and Engine

> In the previous step, you found out that the language host sends requests to the engine to fullfil your to be

Reviewer 3: ```suggestion
In the previous step, you found out that the language host sends requests to the engine to fullfil your to-be
```

infrastructure.

It is now that our Pulumi engine gets to work. The engine combines the intended model of the infrastructure
received from the language host, the current state recorded in the state backend and the actual resource state
to compute which actions need to be executed to bring the actual state in line with the intended model.

Our little example contains dependencies:
every `s3.BucketObject` uses the `bucket.id` as way to define in which bucket these objects should be stored.
The property `id` from the `s3.Bucket` is an `Output`. Outputs are Pulumi's way of tracking which property of one resource

> is required by another, hereby creating a dependency between the resources. Our engine uses all these outputs

Reviewer 3: ```suggestion
is required by another, hereby creating a dependency between the resources. The engine uses all these outputs
```

passed from one resource to another as the vertices in a directed acyclic graph (DAG). The engine determines
the order of actions based on this graph.

> On a first run of `pulumi up` of our example program, our engine will first create the bucket, wait for

Reviewer 3: ```suggestion
On a first run of `pulumi up` of our example program, the engine will first create the bucket, wait for
```

the provisioning to be complete, after which the actual bucket id is passed to the creation of
all the bucket objects. Since none of the bucket objects depend on other resources, these bucket
objects can all be provisioned concurrently. The engine will

> On a second run, assuming no modifications to our example program, the Pulumi engine will compare the to be model

Reviewer 3: ```suggestion
On a second run, assuming no modifications to our example program, the Pulumi engine will compare the to-be model
```

Let's crank up the number of bucket objects to 11.

```python
from pulumi_aws import s3

bucket = s3.Bucket('bucket')

for i in range(11): # <- nr of objects increased by 1
    s3.BucketObject(
        f'object-{i}',
        s3.BucketObjectArgs(
            bucket=bucket.id,
            key=str(i),
        )
    )
```

The third Pulumi run will find the bucket and the first 10 objects in the state. The only action taken
now is the creation of the 11th bucket object. This brings the actual state back in sync with your
intended model.

```sh
$ pulumi up

Updating (<masked>/dev)

View Live: https://app.pulumi.com/<masked>/blog_code/dev/updates/2

     Type                    Name           Status
     pulumi:pulumi:Stack     blog_code-dev
 +   └─ aws:s3:BucketObject  object-10      created

Resources:
    + 1 created
    12 unchanged

Duration: 5s

```

Although you create the intended model of your infrastructure with an imperative language, our engine definitely
processes this in a declarative way.

### Providers

The above examples hopefully made clear that the Pulumi engine calculates a set of actions to declaratively bring

> the actual state in sync with your intended model. But our engine is not the component which knows how to

Reviewer 3: ```suggestion
the actual state in sync with your intended model. But the engine is not the component which knows how to
```

> talk to all the different APIs of cloud and tool vendors. That's the role of the providers.

Reviewer 3: ```suggestion
talk to all the different APIs of cloud and tool vendors. That's the role of the provider.
```

The engine and the provider processes are connected with a gRPC connection. This is similar to the connection
between the language host and the engine, with the only difference of which requests are sent. The API
that providers expose are of a [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) nature:

* Create: create a new resource

> * Check: read information of an existing resource

Reviewer 3: Maybe swap this to align with the R in CRUD just cited?

```suggestion
* Read: check information of an existing resource
```

* Update: update an existing resource with modified information
* Delete: delete an existing resource when no longer needed

The providers receive requests from the engine dependent on the set of actions the engine calculated.
A provider doesn't know anything about the state, the correlation between resource and so forth.

The nature of the provider API is clearly imperative.

### Summary

If the question ever pops up again whether Pulumi is declarative or imperative, the answer is clearly we are both.
It is only based on which component of our architecture you are talking about:

> * Language host: imperative & declarative

Reviewer 3: Should we mention program code here as well, since that's what generally leads to the confusion? The code that we write being itself declarative (even though written in an imperative language)?

Author: @cnunciato the reason I put both `imperative` and `declarative` is because we support languages in both categories:

- imperative: JS/TS, Go, Python, ...
- declarative: YAML

If that wasn't clear when you read this, do you know how to clarify this?

Reviewer 3: I'd love it if we could just say "declarative", since `new s3.Bucket("foo")` is as declarative a statement as `{ foo: { type: "aws:s3:Bucket" } }` is -- but if that's too much, maybe this?

```suggestion
* Language host (your program code): declarative and imperative
```

Reviewer 3: Fair enough, and I defer to you of course. To me though, this is exactly where things get squishy and lead to confusion: _Look -- it's imperative because it's JavaScript!_.  But`var foo = new Foo()` is a declaration -- it's a declaration in an imperative language, yes, but the expression itself is declarative in the same way Adam's `make poop` reference is. IMO. 😇
